Tensile/SolutionStructs.py

```diff
@@ -216,37 +244,181 @@ def __init__(self, problemTypeOut, convolutionType, config):
+      else:
+        raise RuntimeError ("backward weights only supports packSpatialDims")
+
+      spatialDims = []
+      # reverse dims  so can pass spatialDims to register functions in 'convolution' order
+      for si,sdim in enumerate(sdims):
+        spatialDims.insert(0, (sidxStart+si, Fbs.Sum, sdim))
+      self.numSpatialDims = len(spatialDims) # dims actually used in the tensor
+
+
       if self.tensorAFormat in ("NCHW", "NCDHW"):
-        self.registerA( [(nidx,Fbs.Batch,ndim), (cidx,Fbs.Sum,cdim)] + spatialDims + filterDims) # TODO, check)
-        self.registerB( [(nidx,Fbs.Batch,ndim), (cidx,Fbs.Sum,cdim)] + spatialDims + filterDims) # TODO, check?
+        self.registerA( [(nidx,Fbs.Sum,ndim), (cidx,Fbs.Free,cdim)] + filterDims + spatialDims)
+        self.registerB( [(nidx,Fbs.Sum,ndim), (kidx,Fbs.Free,kdim)] + spatialDims)
       else:
-        raise RuntimeError ("unknown activation format '%s'"%self.tensorAFormat)
-      problemTypeOut["NumIndicesC"] = 2
+        raise RuntimeError ("unknown tensorA format '%s'"%self.tensorAFormat)
 
-    problemTypeOut["IndexAssignmentsA"] = [x[0] for x in self.indexA[::-1]]
-    problemTypeOut["IndexAssignmentsB"] = [x[0] for x in self.indexB[::-1]]
+      problemTypeOut["NumIndicesC"] = 2+len(filterDims)
 
+    # convert from convolution order to tensor order:
+    self.indexA.reverse()
+    self.indexB.reverse()
+
+    problemTypeOut["IndexAssignmentsA"] = [x[0] for x in self.indexA]
+    problemTypeOut["IndexAssignmentsB"] = [x[0] for x in self.indexB]
+    problemTypeOut["UseBeta"] = False # MI kernels don't use beta
 
     problemTypeOut["SetConstStrideA"]=[]
     for (idx,fbs,dim) in self.indexA:
       if dim.strideA != -1:
         problemTypeOut["SetConstStrideA"].append([idx,dim.strideA])
+    problemTypeOut["SetConstStrideA"].sort()
 
     problemTypeOut["SetConstStrideB"]=[]
     for (idx,fbs,dim) in self.indexB:
       if dim.strideB != -1:
         problemTypeOut["SetConstStrideB"].append([idx,dim.strideB])
+    problemTypeOut["SetConstStrideB"].sort()
 
-    if 0 in [x[0] for x in problemTypeOut["SetConstStrideA"]] or \
-       0 in [x[0] for x in problemTypeOut["SetConstStrideB"]]:
-     print ("warning: need UseInitialStrides")
-     #problemTypeOut["UseInitialStrides"] = 1
-   #else:
-     #problemTypeOut["UseInitialStrides"] = 0
-
+    if [x for x in problemTypeOut["SetConstStrideA"] if x==[0,1]] or \
+       [x for x in problemTypeOut["SetConstStrideB"] if x==[0,1]]:
+      problemTypeOut["UseInitialStrides"] = False
+    else:
+      problemTypeOut["UseInitialStrides"] = True
 
     #self.printUsage(problemTypeOut)
 
+  def dimIdx(self, convolutionChar):
+    return self.convolutionDims[convolutionChar][0]
+
+  def convolutionChar(self, dimIdx):
+    return self.indexAssignments[dimIdx][1].shortChar
+
+  def makeProblem(self, keepTbd, n, c, k, spatialIn=None):
+    """
+    Generate valid problem dims
```

Comment:
Seems like we should be able to further split this into testable functions.

Also, have you looked at how TensorFlow deals with placeholder values?  Could be useful.